### PR TITLE
chore: Create a patching script for sibling deps

### PR DESCRIPTION
**Pull request by IvayloAndonov**
# Context
As a continuation of **ADP-2700**: Investigate and implement wildcard versions for sibling dependencies
 
We probably need to create a script that patches all package.json files before publishing. (replacing sibling deps version from `workspace:~` notation to the server versions extracted from package JSON files )

# Proposed Solution

**Usage** (root)

`yarn patch-sibling-deps-versions`

Screenshot 2023-05-01 at 16 40 51

# Important Changes Introduced
- create a patching script that updates a package version of each sibling dependency
- use `workspace:~` notation for local dependency resolution

### CODE REVIEW

File: packages/cardano-services-client/package.json [hunk 1]
```diff
@@ -53,7 +53,7 @@
     "test:debug": "DEBUG=true yarn test"
   },
   "devDependencies": {
-    "@cardano-sdk/util-dev": "^0.8.0",
+    "@cardano-sdk/util-dev": "workspace:*",
```
Comment (mkazlauskas) [inline packages/cardano-services-client/package.json:56]: Tilde range matches our pre-1.0 semver strategy
```suggestion
    "@cardano-sdk/util-dev": "workspace:~",
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/projection-typeorm/package.json [hunk 2]
```diff
@@ -62,7 +62,7 @@
     "uuid": "^9.0.0"
   },
   "devDependencies": {
-    "@cardano-sdk/util-dev": "^0.8.0",
+    "@cardano-sdk/util-dev": "0.8.0",
```
Comment (rhyslbw) [inline packages/projection-typeorm/package.json:65]:
```suggestion
    "@cardano-sdk/util-dev": "workspace:~",
```

File: scripts/patch-versions/util.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    acc.then(obj => obj[file.name] = file.version);
+    return acc;
+}, Promise.resolve({}));
+
+const updatePackageJson = (filesData, packageVersionsMap) => {
+    const { packageJsonPath, packageJsonData } = filesData;
+    const dep = packageJsonData.dependencies || [];
+    const devDep = packageJsonData.devDependencies || [];
+
+    console.log('Package JSON file path: ', packageJsonPath)
```
Comment (rhyslbw) [inline scripts/patch-versions/util.js:26]:
```suggestion
    console.log(`Updating ${packageJsonPath}`)
```
Is this a string?

File: packages/projection-typeorm/package.json (hunk 2 above)
Comment (mkazlauskas) [inline packages/projection-typeorm/package.json:65]:
```suggestion
    "@cardano-sdk/util-dev": "workspace:~",
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/projection-typeorm/package.json (hunk 2 above)
Comment (IvayloAndonov) [reply packages/projection-typeorm/package.json]: f169a8a5f2791b39243b3f8918db17305b909da4

File: scripts/patch-versions/util.js (hunk 3 above)
Comment (IvayloAndonov) [reply scripts/patch-versions/util.js]: f169a8a5f2791b39243b3f8918db17305b909da4

**Review by rhyslbw — DISMISSED** [review]
Great work @IvayloAndonov :muscle:

File: scripts/patch-versions/util.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+const fs = require('fs');
+const path = require("path");
+
+const getDirsFromPath = (dirsPath) => fs.readdirSync(dirsPath).filter(dirName => isDirectory(dirsPath, dirName)).map(dirName => path.resolve(dirsPath, dirName));
+
+const isDirectory = (dirPath, fileOrDirName) => fs.statSync(path.join(dirPath, fileOrDirName)).isDirectory();
+
+const hasPackageJSONFile = (dirPath) => fs.existsSync(path.join(dirPath, 'package.json'));
+
+const readFilePromise = (file) => new Promise((resolve, reject) => fs.readFile(file, 'utf-8', (err, data) => err ? reject(err) : resolve(data)));
```
[the remark this answers is not in the document]
Comment (IvayloAndonov) [reply scripts/patch-versions/util.js]: Yeah, this way is more simplified, updated to `readFileSync`

**Comment by IvayloAndonov** [comment]
We verified that Lerna and Yarn would patch package.json during publishing opaquely, not needed.

**Review by iccicci — COMMENTED** [review]
LGTM

**Pull request closed without merging**